add observability topology for PowerScale, UT, e2e and etc

YianZong opened the pull request:
# Description
add observability topology for PowerScale, UT, e2e and etc

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/488 |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] UT
Current coverage:
`[root@vpi6179 controllers]# go test ./... -coverprofile cover.out
ok      github.com/dell/csm-operator/controllers        0.591s  coverage: 88.5% of statements
[root@vpi6179 modules]# go test ./... -coverprofile cover.out
ok      github.com/dell/csm-operator/pkg/modules        0.105s  coverage: 88.9% of statements
`
- [x] E2E
`Running Suite: CSM Operator End-to-End Tests
============================================
Random Seed: 1664430799 - Will randomize all specs
Will run 1 of 1 specs

STEP: Getting test environment variables
STEP: Reading values file
STEP: Getting a k8s client
[run-e2e-test]E2E Testing
  Running all test Given Test Scenarios
  /home/dellemc/csm-operator/tests/e2e/e2e_test.go:75
[It] Running all test Given Test Scenarios
  /home/dellemc/csm-operator/tests/e2e/e2e_test.go:75
STEP: Starting: Install PowerScale Driver(With Observability)
STEP:      Executing  Given an environment with k8s or openshift, and CSM operator installed
STEP:      Executing  Apply custom resources
Sep 29 01:53:23.261: INFO: Running '/usr/bin/kubectl --namespace=test-isilon apply --validate=true -f -'
Sep 29 01:53:23.507: INFO: stderr: ""
Sep 29 01:53:23.507: INFO: stdout: "containerstoragemodule.storage.dell.com/test-isilon created\n"
STEP:      Executing  Validate custom resources
STEP:      Executing  Validate [powerscale] driver is installed
STEP:      Executing  Validate [observability] module is installed
STEP:      Executing  Enable forceRemoveDriver on CR
STEP:      Executing  Delete resources
STEP: Ending: Install PowerScale Driver(With Observability)
STEP:
STEP:
STEP: Starting: Install PowerScale Driver(Standalone), Enable Observability
STEP:      Executing  Given an environment with k8s or openshift, and CSM operator installed
STEP:      Executing  Apply custom resources
Sep 29 01:53:38.553: INFO: Running '/usr/bin/kubectl --namespace=test-isilon apply --validate=true -f -'
Sep 29 01:53:38.802: INFO: stderr: ""
Sep 29 01:53:38.802: INFO: stdout: "containerstoragemodule.storage.dell.com/test-isilon created\n"
STEP:      Executing  Validate custom resources
STEP:      Executing  Validate [powerscale] driver is installed
STEP:      Executing  Validate [observability] module is not installed
STEP:      Executing  Enable [observability] module
STEP:      Executing  Validate [powerscale] driver is installed
STEP:      Executing  Validate [observability] module is installed
STEP:      Executing  Enable forceRemoveDriver on CR
STEP:      Executing  Delete resources
STEP: Ending: Install PowerScale Driver(Standalone), Enable Observability
STEP:
STEP:
STEP: Starting: Install PowerScale Driver(With Observability), Disable Observability module"
STEP:      Executing  Given an environment with k8s or openshift, and CSM operator installed
STEP:      Executing  Apply custom resources
Sep 29 01:53:53.878: INFO: Running '/usr/bin/kubectl --namespace=test-isilon apply --validate=true -f -'
Sep 29 01:53:54.166: INFO: stderr: ""
Sep 29 01:53:54.166: INFO: stdout: "containerstoragemodule.storage.dell.com/test-isilon created\n"
STEP:      Executing  Validate custom resources
STEP:      Executing  Validate [powerscale] driver is installed
STEP:      Executing  Validate [observability] module is installed
STEP:      Executing  Disable [observability] module
STEP:      Executing  Validate [powerscale] driver is installed
STEP:      Executing  Validate [observability] module is not installed
STEP:      Executing  Enable forceRemoveDriver on CR
STEP:      Executing  Delete resources
STEP: Ending: Install PowerScale Driver(With Observability), Disable Observability module"
STEP:
STEP:

• [SLOW TEST:55.986 seconds]
[run-e2e-test]E2E Testing
/home/dellemc/csm-operator/tests/e2e/e2e_test.go:74
  Running all test Given Test Scenarios
  /home/dellemc/csm-operator/tests/e2e/e2e_test.go:75
------------------------------

Ran 1 of 1 Specs in 56.344 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 59.587013979s
Test Suite Passed

`
- [x] Manual
![image](https://user-images.githubusercontent.com/63342088/192953433-edeabb87-1e85-4711-b8d5-502ebf3fd720.png)

rajendraindukuri commented:
Hi @YianZong Please do not merge this PR as posting activities are in progress for CSM Operator. Once those are done, we can merge. Blocking this PR for now.

forrestxia commented on operatorconfig/moduleconfig/common/version-values.yaml:
```diff
@@ -10,3 +10,4 @@ powerscale:
   v2.4.0:
       authorization: "v1.4.0"
       replication: "v1.3.0"
+      observability: "v1.3.0"
```
different obs modules have different versions, shall we use more fine-graded key name? e.g. obs-topology?

YianZong replied:
@forrestxia this is observability module version 1.3.0, rather than individual component image version

forrestxia replied:
observability module version is an abstract concept-only version. If so, shall we use v1.4.0 for the coming release? @YianZong

YianZong replied:
@forrestxia We add v1.3.0 first, then add v1.4.0 for Q4 release.

forrestxia replied:
ok, see. thanks for explanation.

YianZong commented:
> Hi @YianZong Please do not merge this PR as posting activities are in progress for CSM Operator. Once those are done, we can merge. Blocking this PR for now.

Hi @rajendraindukuri, when will the posting be done?

rajendraindukuri commented:
> > Hi @YianZong Please do not merge this PR as posting activities are in progress for CSM Operator. Once those are done, we can merge. Blocking this PR for now.
> 
> Hi @rajendraindukuri, when will the posting be done?

> Hi @YianZong Since RTS was done yesterday. We started posting today.. If all goes good, should be done within a week. Will keep you posted

taohe1012 commented on pkg/modules/observability.go:
```diff
@@ -0,0 +1,141 @@
+	if err != nil {
+		return YamlString, err
+	}
+	// YamlString = utils.ModifyCommonCR(string(buf), cr)
```
this line could be deleted

YianZong replied:
addressed

taohe1012 commented on controllers/csm_controller.go:
```diff
@@ -620,6 +643,19 @@ func (r *ContainerStorageModuleReconciler) SyncCSM(ctx context.Context, cr csmv1
 			}
 		}
 
+		// if Observability is enabled, create or update obs components: topology, metrics of PowerScale and PowerFlex
```
for this block codes, how about Extract them into a function

taohe1012 replied:
because in the future, metrics codes will be added to this block too

YianZong replied:
addressed

taohe1012 commented on pkg/modules/observability.go:
```diff
@@ -0,0 +1,141 @@
+}
+
+// ObservabilityPrecheck  - runs precheck for CSM Observability
+func ObservabilityPrecheck(ctx context.Context, op utils.OperatorConfig, topo csmv1.Module, cr csmv1.ContainerStorageModule, r utils.ReconcileCSM) error {
```
better rename `topo csmv1.Module` to `obs csmv1.Module`

YianZong replied:
addressed

YianZong commented on samples/storage_csm_powerscale_v240.yaml:
```diff
@@ -308,3 +308,22 @@ spec:
           # Allowed values: time
           - name: "RETRY_INTERVAL_MAX"
             value: "5m"
+
```
Observability v1.3.0 aligns with csi-powerscale 2.4

YianZong commented:
Hi Code Owners, please kindly review the PR and let me know if any change needed. Thank you!

forrestxia commented:
@YianZong "Check for forbidden words" seems required, and no result yet, please check what to be done.

YianZong commented:
> @YianZong "Check for forbidden words" seems required, and no result yet, please check what to be done.

@forrestxia The check of "Check for forbidden words" is commented as removed in the actions.yml.

YianZong commented:
Rebased main branch

rensyct commented on api/v1alpha1/types.go:
```diff
@@ -30,6 +30,9 @@ type DriverType string
 type ModuleType string
 
+// OberservabilityComponentType - type representing the type of components inside observability module. e.g. - topology
+type OberservabilityComponentType string
```
spell check Oberservability -> Observability

rensyct commented on api/v1alpha1/types.go:
```diff
@@ -30,6 +30,9 @@ type DriverType string
 // ModuleType - type representing the type of the modules. e.g. - authorization, podmon
 type ModuleType string
 
+// OberservabilityComponentType - type representing the type of components inside observability module. e.g. - topology
```
same comment as below

rensyct commented on api/v1alpha1/types.go:
```diff
@@ -49,6 +52,9 @@ const (
 
+	// Topology - placeholder for constant topology
+	Topology OberservabilityComponentType = "topology"
+
```
Same comment as above

rensyct commented on controllers/csm_controller.go:
```diff
@@ -497,6 +497,24 @@ func (r *ContainerStorageModuleReconciler) oldStandAloneModuleCleanup(ctx contex
+			}
+			for _, cluster := range clusterClients {
+				// remove module observability
+				log.Infow("Deleteing observability")
```
spell check Deleteing -> deleting

rensyct commented on controllers/csm_controller.go:
```diff
@@ -780,6 +821,14 @@ func (r *ContainerStorageModuleReconciler) removeDriver(ctx context.Context, ins
 
+		// remove module observability
+		if observabilityEnabled, _ := utils.IsModuleEnabled(ctx, instance, r, csmv1.Observability); observabilityEnabled {
+			log.Infow("Deleteing observability")
```
Same comment as above

rensyct commented on pkg/modules/testdata/cr_powerscale_observability.yaml:
```diff
@@ -0,0 +1,34 @@
+spec:
+  driver:
+    csiDriverType: "isilon"
+    configVersion: v2.2.0
```
Should this not be 2.4.0?

rensyct commented on pkg/modules/testdata/cr_powerscale_observability.yaml:
```diff
@@ -0,0 +1,34 @@
+    authSecret: test-isilon-creds 
+    replicas: 1
+    common:
+      image: "dellemc/csi-isilon:v2.2.0"
```
Same comment as above

bharathsreekanth commented:
@YianZong v1alpha1 to v1 changes will affect some of the code in here. I personally preferred to have that change go in first, so that all the teams can do the changes needed based on that. Let us discuss this in the meeting today to see how to coordinate.

YianZong commented:
Thanks, @rensyct all addressed.

YianZong commented:
Rebased main branch and updated v1alpha to v1 in newly added files.
Manually verified again:
![image](https://user-images.githubusercontent.com/63342088/195515330-ddad646b-096e-4510-b56f-7af0b6626f9b.png)
